.travis.yml

```diff
@@ -2,33 +2,53 @@ language: python
 python: 3.7
 sudo: true
 dist: bionic
+
+env:
+  global:
+    - DOCKER_CACHE_IMAGE=athenian/api
+    - GOOGLE_DOCKER_IMAGE=gcr.io/athenian-1/api
+
 cache:
-  pip: true
   directories:
-    - $HOME/docker
+    - "$HOME/.cache/pip"
+    - "$HOME/google-cloud-sdk/"
+    - "$HOME/docker"
+
 before_cache:
   - chown -R travis:travis $HOME/.cache/pip
-  - docker save -o $HOME/docker/images.tar athenian/api:latest
+  - docker save -o $HOME/docker/images.tar ${DOCKER_CACHE_IMAGE}:latest
+
+install:
+  - cd server
+  - pip install -r requirements.txt
+  - pip install -r lint-requirements.txt
+  - pip install -r test-requirements.txt
+  - pip install -e .
 
-matrix:
+stages:
+  - name: test
+  - name: deploy
+    if: branch = master AND type != pull_request
+
+jobs:
   fast_finish: true
   include:
-    - install:
-        - cd server
-        - pip install -r requirements.txt
-        - pip install -r lint-requirements.txt
-        - pip install -r test-requirements.txt
-        - pip install -e .
+    - stage: test
+      name: 'Test'
       script:
         - flake8
         - pytest
-    - before_install:
```

Comment:
There should be two parallel testing stage steps: the one that runs flake8 and pytest and the one that builds the image.

The deployment step should simply `make docker-push`.

Rationale: it has been common to screw up the docker build and the failure will show up only after the merge.

Reply:
I think I get the problem: we will run the docker build twice after merge. We could restrict the testing step that builds docker only for PRs.

Reply:
just to make sure that I understood you correctly:
- for a regular branch commit we run only tests
- for a PR we run tests and docker build/docker run in parallel
- for master we run run tests and docker build/docker run in parallel + deploy at the end, am I right?

Reply:
Almost. Change:
* for master we run run tests and deploy at the end (no docker build/run in the testing stage).

Reply:
Done